Added alert component

ettiennelouw opened the pull request:
I added simple alert components with default values for info, success, warning & danger, and tried to cover all the examples on the tailwindUI page for alerts. I use the card components as more or less a template, I don't see any tests for cards so I did not do tests for the alert, should I?

Options:
- padding = string
- rounded = string
- border = bool
- shadow = bool
- set icon = string
- set heading = string
- set text = string
- actions = to insert buttons or something
- set alertClasses="" function to add more custom classes
- set dismiss action with event wire:click="dismiss"

ettiennelouw commented:
`<x-alert border alertClasses="my-2">
                Lorem ipsum dolor sit amet consectetur adipisicing elit. Aliquid pariatur, ipsum similique veniam quo totam eius aperiam dolorum.
            </x-alert>
            <x-alert icon="user" alertClasses="my-2" padding="p-10"/>
            <x-alert info shadow icon="information-circle" heading="Attention information" text="Lorem ipsum dolor sit amet consectetur adipisicing elit. Aliquid pariatur, ipsum similique veniam quo totam eius aperiam dolorum." alertClasses="my-2"/>
            <x-alert warning dismiss icon="exclamation" text="Lorem ipsum dolor sit amet consectetur adipisicing elit. Aliquid pariatur, ipsum similique veniam quo totam eius aperiam dolorum." alertClasses="my-2" />
            <x-alert success icon="information-circle" text="Lorem ipsum dolor sit amet consectetur adipisicing elit. Aliquid pariatur, ipsum similique veniam quo totam eius aperiam dolorum." alertClasses="my-2"/>
            <x-alert danger shadow border dismiss icon="x-circle" heading="Attention danger" text="Lorem ipsum dolor sit amet consectetur adipisicing elit. Aliquid pariatur, ipsum similique veniam quo totam eius aperiam dolorum." alertClasses="my-2">
                <x-slot name="actions">
                    <button type="button" class="bg-red-50 px-2 py-1.5 rounded-md text-sm font-medium text-red-700 hover:bg-red-100 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-offset-green-50 focus:ring-green-600">View status</button>
                    <button type="button" class="ml-3 bg-red-50 px-2 py-1.5 rounded-md text-sm font-medium text-red-700 hover:bg-red-100 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-offset-green-50 focus:ring-green-600">Dismiss</button>
                </x-slot>
            </x-alert>`

PH7-Jack commented:
@ettiennelouw Thanks for contributing. I'm glad about your interest. I'm reviewing this PR and creating the documentation.
Could you add some tests for this component?

ettiennelouw commented:
Tests added if you can just verify and see if they are sufficient. Also made the icon be added by default if you choose info, warning, success and danger but you can override them by specifying a icon=""

PH7-Jack reviewed (changes requested):
Could you refactor the Alert class as I commented in the avatar PR?

Could you create different components for each alert type?
Create a base alert and extends the base class to create other classes.
Small components are simple to maintain

PH7-Jack commented on src/View/Components/Alert.php:
```diff
@@ -0,0 +1,142 @@
+        bool $info = false,
+        bool $warning = false,
+        bool $success = false,
+        bool $danger = false,
```
```suggestion
        bool $error = false,
```

ettiennelouw replied:
Updated it to be more in line with the buttons
public bool $info = false,
public bool $warning = false,
public bool $positive = false,
public bool $negative = false,

PH7-Jack commented on src/View/Components/Alert.php:
```diff
@@ -0,0 +1,142 @@
+        bool $warning = false,
+        bool $success = false,
+        bool $danger = false,
+        bool $dismiss = false,
```
```suggestion
        bool $dismissable = false,
```

PH7-Jack commented on src/View/Components/Alert.php:
```diff
@@ -0,0 +1,142 @@
+        ?string $padding = 'p-4',
+        ?string $rounded = 'rounded-md',
+        ?string $icon = null,
+        ?string $heading = null,
+        ?string $text = null,
```
```suggestion
        ?string $title = null,
        ?string $message = null,
```

PH7-Jack commented on resources/views/components/alert.blade.php:
```diff
@@ -0,0 +1,32 @@
+            @endif
+        </div>
+        @if($dismiss)
+            <div class="ml-auto pl-3">
+                <div class="-mx-1.5 -my-1.5">
+                    <button type="button" wire:click="dismiss" class="inline-flex {{ $backgroundColor }} rounded-md p-1.5 {{ $subjectColor }} focus:outline-none focus:ring-2 focus:ring-offset-2">
+                        <span class="sr-only">Dismiss</span>
+                        <x-icon name="x" class="h-5 w-5" />
+                    </button>
+                </div>
+            </div>
```
Add a **wire:click** with a default action is bad for the reusability, many components can have different methods to handle the dismissal of an alert. Just a slot will be good, if the alert is dismissable a custom logic will be required to the Livewire Component to show/hide the alert.

ettiennelouw commented:
I updated the alerts and tests, split out the class to alert and basealert classes just not sure if it is worth it to split each into its own wont that create a lot of duplicates? Or I'm not 100% sure i understand :)

_Could you create different components for each alert type?
Create a base alert and extends the base class to create other classes.
Small components are simple to maintain_

PH7-Jack commented:
> I updated the alerts and tests, split out the class to alert and basealert classes just not sure if it is worth it to split each into its own wont that create a lot of duplicates? Or I'm not 100% sure i understand :)
> 
> _Could you create different components for each alert type? Create a base alert and extends the base class to create other classes. Small components are simple to maintain_

You can continue using the same blade file for all alerts, but multiple classes with custom CSS implementation. This way will reduce the maintenance process

```
Components/Alerts/Info.php
Components/Alerts/Success.php
...

```


But it was my old idea, we can now still use one blade and one PHP class, ignore my the past message.
In one class will be possible to have all tailwind colors in the future.

PH7-Jack commented on src/config/wireui.php:
```diff
@@ -59,9 +59,9 @@
         |
      */
     'components' => [
-        'avatar' => [
```
rollback the avatar component

PH7-Jack commented on resources/views/components/alert.blade.php:
```diff
@@ -0,0 +1,32 @@
+<div class="{{ $alertClasses }}">
```
```suggestion
<div {{ $attributes->class($alertClasses) }}>
```

PH7-Jack commented on resources/views/components/alert.blade.php:
```diff
@@ -0,0 +1,32 @@
+<div class="{{ $alertClasses }}">
+    <div class="flex">
+
+        @if($icon)
```
```suggestion
        @if ($icon)
```

PH7-Jack commented on resources/views/components/alert.blade.php:
```diff
@@ -0,0 +1,32 @@
+            </div>
+        @endif
+
+        <div class="ml-3">
+            <h3 class="text-sm font-semibold {{ $subjectColor }}">{{ $title }}</h3>
+            <div class="@if(!is_null($title)) mt-2 @endif text-sm {{ $subjectColor }}">
+                <p>{{ $message ?? $slot }}</p>
+            </div>
+            @if($actions)
+                <div class="mt-4">
+                    <div class="-mx-2 -my-1.5 flex">
+                        {{ $actions }}
+                    </div>
+                </div>
+            @endif
+        </div>
```
```suggestion
        <div class="ml-3">
            @if ($title)
                <h3 class="mb-2 text-sm font-semibold {{ $subjectColor }}">
                    {{ $title }}
                </h3>
            @endif

            <div class="text-sm {{ $subjectColor }}">
                <p>{{ $message ?? $slot }}</p>
            </div>
            
            @if ($actions)
                <div class="mt-4">
                    <div class="-mx-2 -my-1.5 flex">
                        {{ $actions }}
                    </div>
                </div>
            @endif
        </div>
```

PH7-Jack commented on resources/views/components/alert.blade.php:
```diff
@@ -0,0 +1,32 @@
+            @endif
+        </div>
+
+        @if($dismissible)
```
```suggestion
        @if ($dismissible)
```

ettiennelouw commented:
I should not have pulled main into my branch, Im closing this one and will make a new pull request.
